dot/types/digest.go

```diff
@@ -15,11 +16,40 @@ func NewDigestItem() scale.VaryingDataType {
 	return scale.MustNewVaryingDataType(ChangesTrieRootDigest{}, PreRuntimeDigest{}, ConsensusDigest{}, SealDigest{})
 }
 
-// NewDigest returns a new Digest from the given DigestItems
+// NewDigest returns a new Digest as a varying data type slice.
 func NewDigest() scale.VaryingDataTypeSlice {
 	return scale.NewVaryingDataTypeSlice(NewDigestItem())
 }
 
+// DigestToString renders a digest varying data type slice as a string.
+func DigestToString(digest scale.VaryingDataTypeSlice) (s string) {
+	elements := make([]string, 0, len(digest.Types))
+	for _, typ := range digest.Types {
+		vdtValue, err := typ.Value()
+		if err != nil {
+			panic(err)
+		}
+
+		var element string
+		switch impl := vdtValue.(type) {
+		case ChangesTrieRootDigest:
+			element = impl.String()
+		case PreRuntimeDigest:
+			element = impl.String()
+		case ConsensusDigest:
+			element = impl.String()
+		case SealDigest:
+			element = impl.String()
+		default:
+			panic(fmt.Sprintf("digest type not implemented for String: %T", impl))
+		}
```

Comment:
I think almost all `VaryingDataType` implementations implements `String` method we could change the `VaryingDataTypeValue` interface to:

```go
type VaryingDataTypeValue interface {
	fmt.Stringer
	Index() uint
}
```

Then you can just use `elements = append(elements, vdtValue.String())`

Reply:
Great suggestion. It required a bit more work than expected, a bunch of types were missing the String method, but it's done now in #2970 and `DigestToString` is no longer needed (this PR just ends up being a unit test :wink:)